sql/sem/tree, backupccl, *: sanitize URLs during Format

michae2 opened the pull request:
Prior to this change, all SQL statements containing URLs would be formatted with the full URL in cleartext, including any secrets such as keys or passwords. These secrets would sometimes show up in the slow query log or sql audit log. This PR adds new functions `tree.(*FmtCtx).FormatURI`and `tree.(*FmtCtx).FormatURIs` which are designed to sanitize URLs during formatting.

See individual commits for details.

Fixes: CRDB-39710, TREQ-284

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126970)
<!-- Reviewable:end -->

michae2 committed:
tree, parser, backupccl: use FormatURI in ALTER BACKUP

Use `tree.(*FmtCtx).FormatURI` in `ALTER BACKUP`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being written to unredacted logs:
- `ALTER BACKUP`

michae2 committed:
tree, parser, backupccl: use FormatURI in SHOW BACKUP

Use `tree.(*FmtCtx).FormatURI` in `SHOW BACKUP`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being written to unredacted logs:
- `SHOW BACKUP`
- `SHOW BACKUPS`

michae2 committed:
tree: add FormatURI and FormatURIs

Add new functions `tree.(*FmtCtx).FormatURI` and
`tree.(*FmtCtx).FormatURIs` which can be used to safely format one or
more string literals containing URLs. During formatting the URLs are
hidden to prevent leaking sensitive information such as passwords and
keys.

There are some special cases for which we do *not* sanitize the URL:
- if the URL is a placeholder like $1
- if the URL is an empty string or an underscore
- if the URL has already been sanitized (e.g. while building clean AST
  nodes for job details) which is communicated by setting the
  `FmtShowFullURIs` flag
- if we're showing passwords, or already hiding constants

Future commits will call these functions when formatting SQL statements
that could contain a URL.

Epic: None

Release note: None

michae2 committed:
tree, parser, backupccl: use FormatURI in BACKUP and RESTORE

Use `tree.(*FmtCtx).FormatURI` in various backup-related statements.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being written to unredacted logs:
- `BACKUP`
- `RESTORE`
- `CREATE SCHEDULE FOR BACKUP`
- `ALTER BACKUP SCHEDULE`

michae2 committed:
tree, parser, changefeedccl: use FormatURI in CREATE CHANGEFEED

Use `tree.(*FmtCtx).FormatURI` in `CREATE CHANGEFEED`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being written to unredacted logs:
- `CREATE CHANGEFEED`
- `CREATE SCHEDULE FOR CHANGEFEED`

dt commented on pkg/ccl/backupccl/alter_backup_schedule_test.go:
```diff
@@ -128,12 +128,12 @@ INSERT INTO t1 values (1), (10), (100);
 		"PAUSED: Waiting for initial backup to complete",
 		"@daily",
-		"BACKUP TABLE mydb.public.t1 INTO LATEST IN 'nodelocal://1/backup/alter-schedule' WITH OPTIONS (detached)",
+		"BACKUP TABLE mydb.public.t1 INTO LATEST IN '*****' WITH OPTIONS (detached)",
```
Hm, maybe `backupccl.emitAlteredSchedule` needs to pass FmtShowPasswords ?

michae2 replied:
I was nervous to do that, because then I think `encryption_passphrase` might be revealed? But I did add a FmtShowFullURIs flag that could work.

I'm not sure whether this output should be sanitized or not, to be honest. It seems like the fact that we're already hiding `encryption_passphrase` means we should also hide the URL, since it could contain secrets?

michae2 replied:
Oh, maybe I'm misunderstanding how `backupccl.emitAlteredSchedule` works.

dt replied:
Eh, we're showing you the schedule _that you just set_ -- scrubbing the URI completely seems wrong. Sanitizing it may or may not be right, but for the purposes of the bug you are trying to fix -- unsanitized URIs in logs if I understand correctly? -- I think this is orthogonal?

michae2 replied:
Ok, thanks. This is helpful!

michae2 replied:
Should be fixed now.

michae2 committed:
tree, parser, importer: use FormatURI in IMPORT

Use `tree.(*FmtCtx).FormatURI` in `IMPORT`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being written to unredacted logs:
- `IMPORT`

michae2 committed:
tree, parser: use FormatURI in EXPORT

Use `tree.(*FmtCtx).FormatURI` in `EXPORT`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being written to unredacted logs:
- `EXPORT`

michae2 committed:
tree, parser, externalconn: use FormatURI in CREATE EXTERNAL CONNECTION

Use `tree.(*FmtCtx).FormatURI` in `CREATE EXTERNAL CONNECTION`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being written to unredacted logs:
- `CREATE EXTERNAL CONNECTION`

michae2 committed:
tree, parser: use FormatURI in COPY

Use `tree.(*FmtCtx).FormatURI` in `COPY`.

Epic: None

Release note (security update): URLs in the following SQL statements are
now sanitized of any secrets before being written to unredacted logs:
- `COPY`

michae2 commented:
This is a simplified version of https://github.com/cockroachdb/cockroach/pull/126185 which will hopefully be more suitable for backport. It replaces the entire URL with `'*****'` instead of calling `SanitizeExternalStorageURI`, and also avoids all of the AST and parser changes.

dt commented on pkg/sql/sem/tree/format.go:
```diff
@@ -469,6 +480,59 @@ func (ctx *FmtCtx) FormatNameP(s *string) {
+			ctx.FormatNode(n)
+			return
+		}
+		var raw, sanitized string
```
nit: I might call this `elided` or `redacted` since "sanitized" implies mostly preserved with just some bits cleaned up, i.e. the granular sanitization routine, but we're not doing that -- we're just eliding it.

michae2 replied:
Done.

rytaft reviewed (commented):
Very nice! A few questions / nits.

> Reviewed 35 of 35 files at r14, 15 of 15 files at r15, 3 of 3 files at r16, 4 of 4 files at r17, 7 of 7 files at r18, 8 of 8 files at r19, 3 of 3 files at r20, 4 of 4 files at r21, 3 of 3 files at r22, 35 of 35 files at r23, 14 of 14 files at r24, 2 of 2 files at r25, 3 of 3 files at r26, 6 of 6 files at r27, 7 of 7 files at r28, 2 of 2 files at r29, 3 of 3 files at r30, 2 of 2 files at r31, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-:-O1sBvuq0yiZHwRjL6uU:b-m9ezco)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @dt, @michae2, @msbutler, and @xinhaoz)

___
*[`pkg/sql/sem/tree/alter_changefeed.go` line 94 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-O1s1p_64nE9W5INfHDH:-O1s1p_64nE9W5INfHDI:b-yhk1da) ([raw file](https://github.com/cockroachdb/cockroach/blob/48fda5322e388f71c98263902f34c9c8baecaaf8/pkg/sql/sem/tree/alter_changefeed.go#L94)):*
> ```Go
> func (node *AlterChangefeedSetOptions) Format(ctx *FmtCtx) {
> 	ctx.WriteString(" SET ")
> 	// TODO(michae2): sanitize SET sink
> ```

Does this change need to happen as part of this PR?
___
*[`pkg/sql/sem/tree/format.go` line 486 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-O1s9hvO3anE9O_AsfDJ:-O1s9hvO3anE9O_AsfDK:b6mbizs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d6bece23595e488fe18aa9feddd0613995de6a5/pkg/sql/sem/tree/format.go#L486)):*
> ```Go
> func (ctx *FmtCtx) FormatURIs(uris []Expr) {
> 	if len(uris) > 1 {
> 		ctx.WriteString("(")
> ```

I noticed that this caused import logic to change with parentheses. Is this going to cause issues anywhere else? Any reason we need this special logic with parentheses now? And is this well tested?
___
*[`pkg/sql/parser/testdata/import_export` line 4 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-O1sBNVgBLuyxGLRf6Vd:-O1sBNVgBLuyxGLRf6Ve:b79q2ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d6bece23595e488fe18aa9feddd0613995de6a5/pkg/sql/parser/testdata/import_export#L4)):*
> ```
> IMPORT TABLE foo FROM PGDUMPCREATE 'nodelocal://0/foo/bar' WITH temp = 'path/to/temp'
> ----
> IMPORT TABLE foo FROM PGDUMPCREATE '*****' WITH OPTIONS (temp = 'path/to/temp') -- normalized!
> ```

should we be concerned about 'path/to/temp'?


<!-- Sent from Reviewable.io -->

dt commented:
> temp = ...

`temp` hasn't been a valid option to IMPORT for 5+ years. The parser testdata is just showing that you can pass any pair of strings to these untyped string option = string value maps, and we don't do any validation of which keys you pass or what you pass for them until execution of the stmt (at which point you'd get `ERROR: invalid option "temp"`. Fortunately we don't put many URIs in these unstructured option maps.

michae2 reviewed (commented):
TFTRs, @dt and @rytaft! Any objection if I add a cluster setting to bring back the old behavior, default off?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-:-O1sLxf15Vh3Brme9frE:bh3yr2o)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @dt, @msbutler, @rytaft, and @xinhaoz)

___
*[`pkg/sql/sem/tree/alter_changefeed.go` line 94 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-O1s1p_64nE9W5INfHDH:-O1sMAm8AOwseuwGyLDO:b-79kg03) ([raw file](https://github.com/cockroachdb/cockroach/blob/48fda5322e388f71c98263902f34c9c8baecaaf8/pkg/sql/sem/tree/alter_changefeed.go#L94)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Does this change need to happen as part of this PR?
</blockquote></details>

Ah, good catch. Yes. Done.
___
*[`pkg/sql/sem/tree/format.go` line 486 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-O1s9hvO3anE9O_AsfDJ:-O1sMGhx1qlL3UgReXAV:bbuxxs5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d6bece23595e488fe18aa9feddd0613995de6a5/pkg/sql/sem/tree/format.go#L486)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I noticed that this caused import logic to change with parentheses. Is this going to cause issues anywhere else? Any reason we need this special logic with parentheses now? And is this well tested?
</blockquote></details>

Great question. Every field using `FormatURIs` _except_ `Import.Files` comes from [string_or_placeholder_opt_list](https://github.com/cockroachdb/cockroach/blob/6f2da79100d4fad647a54fee0c7a52ad0ca74827/pkg/sql/parser/sql.y#L3848-L3856) in the parser, which makes the parentheses optional for a single value, so I made that the default behavior. This matches the behavior of [StringOrPlaceholderOptList.Format](https://github.com/cockroachdb/cockroach/blob/6f2da79100d4fad647a54fee0c7a52ad0ca74827/pkg/sql/sem/tree/backup.go#L250-L258).

`Import.Files` is a special case: it comes from [string_or_placeholder_list surrounded by parentheses](https://github.com/cockroachdb/cockroach/blob/6f2da79100d4fad647a54fee0c7a52ad0ca74827/pkg/sql/parser/sql.y#L4045).

This is tested by the cases in `pkg/sql/parser/testdata/import_export` which shouldn't have changed w.r.t. parentheses.

(The other weird one is the `INCREMENTAL FROM` clause of BACKUP which comes from [string_or_placeholder_list without any parentheses](https://github.com/cockroachdb/cockroach/blob/6f2da79100d4fad647a54fee0c7a52ad0ca74827/pkg/sql/parser/sql.y#L4100), so I also had to add a special case for that.)


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> Any objection if I add a cluster setting to bring back the old behavior, default off?

Actually, this is turning out to be very painful to plumb into all calls to AsString*. Forget I said this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-:-O1sUw3fCVkFTeKFBMt4:b-1wn128)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @dt, @msbutler, @rytaft, and @xinhaoz)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 36 of 36 files at r37, 14 of 14 files at r38, 2 of 2 files at r39, 3 of 3 files at r40, 8 of 8 files at r41, 7 of 7 files at r42, 2 of 2 files at r43, 3 of 3 files at r44, 2 of 2 files at r45, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-:-O1y9pLrCRaGnz4akrRY:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andyyang890, @dt, @msbutler, and @xinhaoz)

___
*[`pkg/sql/sem/tree/format.go` line 486 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/126970#-O1s9hvO3anE9O_AsfDJ:-O1y9jSZ7r9aEh0Pz9FI:b-wsvasg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d6bece23595e488fe18aa9feddd0613995de6a5/pkg/sql/sem/tree/format.go#L486)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Great question. Every field using `FormatURIs` _except_ `Import.Files` comes from [string_or_placeholder_opt_list](https://github.com/cockroachdb/cockroach/blob/6f2da79100d4fad647a54fee0c7a52ad0ca74827/pkg/sql/parser/sql.y#L3848-L3856) in the parser, which makes the parentheses optional for a single value, so I made that the default behavior. This matches the behavior of [StringOrPlaceholderOptList.Format](https://github.com/cockroachdb/cockroach/blob/6f2da79100d4fad647a54fee0c7a52ad0ca74827/pkg/sql/sem/tree/backup.go#L250-L258).

`Import.Files` is a special case: it comes from [string_or_placeholder_list surrounded by parentheses](https://github.com/cockroachdb/cockroach/blob/6f2da79100d4fad647a54fee0c7a52ad0ca74827/pkg/sql/parser/sql.y#L4045).

This is tested by the cases in `pkg/sql/parser/testdata/import_export` which shouldn't have changed w.r.t. parentheses.

(The other weird one is the `INCREMENTAL FROM` clause of BACKUP which comes from [string_or_placeholder_list without any parentheses](https://github.com/cockroachdb/cockroach/blob/6f2da79100d4fad647a54fee0c7a52ad0ca74827/pkg/sql/parser/sql.y#L4100), so I also had to add a special case for that.)
</blockquote></details>

Thanks for the explanation!


<!-- Sent from Reviewable.io -->

michae2 commented:
Thanks for the reviews!

bors r=dt,rytaft
